# avantonder/assembleBAC-ONT: Output

## Introduction

This document describes the output produced by the `assembleBAC-ONT` pipeline. The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:
* [Aggregate pre-demultiplexed reads from MinKNOW/Guppy](#nanopore-artic-guppyplex)
* [Filter long reads](#filter-long-reads) 
* [Read subsampling](#read-subsampling)
* [Remove adapter sequences](#remove-adapter-sequences)
* [*de novo* genome assembly](#de-novo-genome-assembly)
* [Sequence Type assignment](#sequence-type-assignment)
* [Genome annotation](#genome-annotation)
* [Genome completeness](#genome-completeness)
* [Assembly metrics](#assembly-metrics) 
* [MultiQC](#multiqc)
* [Pipeline information](#pipeline-information)

### Nanopore: artic guppyplex

[artic guppyplex](https://artic.readthedocs.io/en/latest/commands/) is used to perform length filtering of the demultiplexed Nanopore reads obtained per barcode.

<details markdown="1">
<summary>Output files</summary>

- `guppyplex/`
  - `*.fastq.gz` files generated by aggregate pre-demultiplexed reads from MinKNOW/Guppy. These files are not saved by default but can be via a custom config file such as the one below.

</details>

### Filter long reads

[Filtlong](https://github.com/rrwick/Filtlong) is used to filter short reads from Oxford Nanopore FASTQ files.

<details markdown="1">
<summary>Output files</summary>

- `filtlong/`
  - `*.log` log files generated by Filtlong.

</details>

### Read Subsampling

[rasusa](https://github.com/mbhall88/rasusa) is used to subsample reads to a depth cutoff of a default of 100.

<details markdown="1">
<summary>Output files</summary>

* `rasusa/`
    * `*.fastq.gz` subsampled fastq files

</details>

### Remove adapter sequences

([`Porechop`](https://github.com/rrwick/Porechop)) is used to remove adapter sequences from Oxford Nanopore reads.

<details markdown="1">
<summary>Output files</summary>

* `porechop/`
    * `*.trim.fastq.gz` adapter trimmed fastq files

</details>

### *de novo* genome assembly

[Flye](https://github.com/fenderglass/Flye) assembles bacterial isolate genomes from Oxford Nanopore reads.

<details markdown="1">
<summary>Output files</summary>

- `flye/`
  - `*_log`: Flye log files

</details>

### Sequence Type assignment

[mlst](https://github.com/tseemann/mlst) is used to assign Sequence Types (STs) to bacterial isolate genomes.

<details markdown="1">
<summary>Output files</summary>

- `mlst/`
  - `*.tsv`: MLST calls in tsv format
- `metadata/`
  - `mlst_summary.tsv`: mlst summary in tsv format

</details>

### Genome annotation

[Bakta](https://github.com/oschwengers/bakta) is used for the annotation of bacterial genomes (isolates, MAGs) and plasmids.

<details markdown="1">
<summary>Output files</summary>

- `bakta/`
  - `*.gff3`: Annotations & sequences in GFF3 format

</details>

### Genome completeness

[CheckM2](https://github.com/chklovski/CheckM2) is used for the rapid assessment of genome bin quality using machine learning

<details markdown="1">
<summary>Output files</summary>

- `checkm2/`
  - `*.tsv`: CheckM2 output in tsv format
- `metadata/`
  - `checkm2_summary.tsv`: CheckM2 summary in tsv format

</details>

### Assembly metrics

[Quast](https://quast.sourceforge.net/) is a quality assessment tool for genome assemblies.

<details markdown="1">
<summary>Output files</summary>

- `metadata/`
  - `transposed_report.tsv`: Quast summary in tsv format

</details>

### MultiQC

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
